### PR TITLE
1.2.0

### PR DESCRIPTION
**Pull request by Konicai**
This is currently only for monitoring changes and progress.

- [x] Implement papi in as many viewable text pieces as possible
- [x] Implement button results running commands as players - I think form specific permissions can come at another time
- [x] Help command - needs to be tested
- [x] Player command to open a form
- [x] Implement a command to make another player open a form
- [x] Split the menu configuration sections into another file - when other pr is merged
- [x] Extra forms for bedrock players
- [x] Extra forms for java players
- [x] On join teleporter
- [x] Display git info on startup

### CODE REVIEW

**Comment by Jens-Co** [comment]
What we can do is make an new join item "Menu" where it opens a new form + GUI with command items/buttons

**Comment by Jens-Co** [comment]
> What we can do is make an new join item "Menu" where it opens a new form + GUI with command items/buttons

With an other config doc "menu.yml"

File: pom.xml [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
     <groupId>dev.projectg</groupId>
     <artifactId>GeyserHub</artifactId>
-    <version>1.1.0-SNAPSHOT</version>
+    <version>1.2.0-SNAPSHOT</version>
```
Comment (Konicai) [inline pom.xml]: remove snapshot

File: src/main/java/dev/projectg/geyserhub/GeyserHubMain.java [hunk 2]
```diff
@@ -1,151 +1,103 @@
...
-            logger.severe("Disabling due to configuration error. Fix the formatting or regenerate a new one");
+        // getting the logger forces the config to load before our loadConfiguration() is called...
+        SelectorLogger logger = SelectorLogger.getLogger();
+
+        try {
+            Properties gitProperties = new Properties();
+            gitProperties.load(FileUtils.getResource("git.properties"));
+            logger.info("Branch: " + gitProperties.getProperty("git.branch", "Unknown") + ", Commit: " + gitProperties.getProperty("git.commit.id.abbrev", "Unknown"));
+        } catch (IOException e) {
+            logger.warn("Unable to load resource: git.properties");
```
Comment (Konicai) [inline src/main/java/dev/projectg/geyserhub/GeyserHubMain.java:43]: print exception message here, full message if debug

Comment (Konicai) [reply src/main/java/dev/projectg/geyserhub/GeyserHubMain.java:43]: Config is loaded after this... maybe forget the latter

File: src/main/java/dev/projectg/geyserhub/GeyserHubMain.java [hunk 3]
```diff
@@ -1,151 +1,103 @@
...
-        if (this.getServer().getPluginManager().getPlugin("Essentials") == null) {
-            Placeholders.essentials = 0;
-        } else {
-            this.setupEconomy();
-        }
-        this.Scheduler();
-        Bukkit.getServer().getPluginManager().registerEvents(new ScoreboardManager(), this);
+        // The random interval broadcast module
+        Broadcast.startBroadcastTimer(getServer().getScheduler());
     }
```
Comment (Konicai) [inline src/main/java/dev/projectg/geyserhub/GeyserHubMain.java:87]: Maybe register listeners in the same order that they are listed in the module

File: src/main/java/dev/projectg/geyserhub/GeyserHubMain.java [hunk 4]
```diff
@@ -1,151 +1,103 @@
...
-            Placeholders.economy = (Economy) economyProvider.getProvider();
-        }
-    }
-
-    private void setupPermissions() {
-        RegisteredServiceProvider permissionProvider = this.getServer().getServicesManager().getRegistration(Permission.class);
-        if (permissionProvider != null) {
-            Placeholders.permission = (Permission) permissionProvider.getProvider();
-        }
+        }, 20L, ScoreboardManager.REFRESH_RATE * 20L);
```
Comment (Konicai) [inline src/main/java/dev/projectg/geyserhub/GeyserHubMain.java:96]: Start schedulers for different modules regardless of config (and handle config within), or only start schedule if enabled in the config

File: src/main/java/dev/projectg/geyserhub/command/GeyserHubCommand.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+            default:
+                sendMessage(commandSender, SelectorLogger.Level.SEVERE, UNKNOWN);
+                break;
+        }
+        return true;
+    }
+
+    private void sendHelp(CommandSender commandSender) {
+        // todo: only show players with the given permissions certain entries? not sure if it can be integrated any way into spigot command completions
+        // todo: check if these are sent on consecutive lines or the same one :(
```
Comment (Konicai) [inline src/main/java/dev/projectg/geyserhub/command/GeyserHubCommand.java]: Make sure help command works. Make sure all commands work.

File: src/main/java/dev/projectg/geyserhub/config/ConfigId.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+    /**
+     * @param fileName the filename, including the extension, of the configuration
+     * @param version the version of the configuration
+     */
+    ConfigId(String fileName, int version) {
+        this.fileName = fileName;
+        this.version = version;
+    }
+}
```
Comment (Konicai) [inline src/main/java/dev/projectg/geyserhub/config/ConfigId.java:25]: Maybe load configs on enum init? can wait

File: src/main/java/dev/projectg/geyserhub/module/menu/MenuUtils.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        if (buttonData.contains("Server") && buttonData.isString("Server")) {
+            String serverName = Objects.requireNonNull(buttonData.getString("Server"));
+            logger.debug(getParentName(buttonData) + "." + buttonData.getName() + " contains BungeeCord target server: " + serverName);
+            return serverName;
+        }
+        return null;
+    }
+
+    @Nonnull
+    public static String getParentName(@Nonnull ConfigurationSection configSection) {
```
Comment (Konicai) [inline src/main/java/dev/projectg/geyserhub/module/menu/MenuUtils.java]: javadoc

File: src/main/java/dev/projectg/geyserhub/module/menu/bedrock/BedrockForm.java [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+
+    private final SelectorLogger logger;
+
+    public final boolean isEnabled;
+    public final String formName;
+
+    private String title;
+    private String content;
+    private List<BedrockButton> allButtons;
+
```
Comment (Konicai) [inline src/main/java/dev/projectg/geyserhub/module/menu/bedrock/BedrockForm.java:51]: have another constructor with formName/title/content/buttons as args. can wait.

File: src/main/java/dev/projectg/geyserhub/module/menu/java/JavaMenu.java [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+
+    private final SelectorLogger logger;
+
+    public final boolean isEnabled;
+    @Nonnull private final String menuName;
+
+    private String title;
+    private int size;
+    private Map<Integer, ItemButton> buttons;
+
```
Comment (Konicai) [inline src/main/java/dev/projectg/geyserhub/module/menu/java/JavaMenu.java:57]: have another constructor with formName/title/content/buttons as args. can wait.

File: src/main/java/dev/projectg/geyserhub/module/world/WorldSettings.java [hunk 10]
```diff
@@ -15,23 +16,22 @@
...
 import org.bukkit.inventory.ItemStack;
 
-
 public class WorldSettings implements Listener {
 
     @EventHandler
     public void onEntityDamage(EntityDamageEvent event) {
+        FileConfiguration config = GeyserHubMain.getInstance().getConfigManager().getFileConfiguration(ConfigId.MAIN);
         if (!(event.getEntity() instanceof Player)) return;
-        Player player = (Player) event.getEntity();
```
Comment (Konicai) [inline src/main/java/dev/projectg/geyserhub/module/world/WorldSettings.java:24]: bracket syntax

File: src/main/resources/selector.yml [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+      Content: "Click on a minigame of choice."
+      Buttons:
+        spleef:
+          Button-Text: "Spleef"
+          ImageURL: "https://www.digminecraft.com/materials/images/snowball.png"
+          Server: "spleef"
+
+        hideseek:
+          Button-Text: "Hide & Seek"
+          Server: "hideseek"
```
Comment (Konicai) [inline src/main/resources/selector.yml:137]: remove any `execute as`  usage here

**Review by Konicai — COMMENTED** [review]
1. New/changed features need to be tested 2. This stuff needs to be addressed:

**Review by Jens-Co — DISMISSED** [review]
All looking good.

**Comment by Konicai** [comment]
still gotta finish handling java menu clicks, then testing

**Comment by Konicai** [comment]
Testing should be done now

**Comment by Konicai** [comment]
Tested a fair amount by me. Forms and menus seem to work great, fixed a fair amount of bugs in this PR's code. Teleporter works fine, didn't really look into broadcast/scoreboard but it looked to work fine.

**Comment by Jens-Co** [comment]
> Tested a fair amount by me. Forms and menus seem to work great, fixed a fair amount of bugs in this PR's code. Teleporter works fine, didn't really look into broadcast/scoreboard but it looked to work fine.

Broadcasts and scoreboards are fine afaik, did some good amount of testing on them.

**Review by Jens-Co — APPROVED** [review]
Guuwd

**Comment by Konicai** [comment]
access item is duplicated if the player is in creative mode 😄

**Pull request merged**